### ISSUE TITLE
v1.0dev cannot parse DataStructure ECOFIN_DSD from IMF, while v0.9 can

### ISSUE THREAD

**Issue by dr-leo** [dr-leo/pandaSDMX#165]
See the following code run with v1.0dev first, then v0.9. v1.0dev raises an XML parse error. Should be fixed pre v1.0 release, #145 

```
In [36]: dsd.id
Out[36]: 'ECOFIN_DSD'

In [37]: dsd_resp = imf.datastructure(dsd.id)
---------------------------------------------------------------------------
XMLParseError                             Traceback (most recent call last)
<ipython-input-37-f737e4f19e00> in <module>
----> 1 dsd_resp = imf.datastructure(dsd.id)

~\Documents\GitHub\pandaSDMX\pandasdmx\api.py in get(self, resource_type, resource_id, tofile, use_cache, dry_run, **kwargs)
    409
    410         # Parse the message
--> 411         msg = reader.read_message(response_content)
    412
    413         # Store the HTTP response with the message

~\Documents\GitHub\pandaSDMX\pandasdmx\reader\sdmxml.py in read_message(self, source, dsd)
    285
    286         # Parse the tree
--> 287         values = self._parse(root)
    288
    289         # Instantiate the message object

~\Documents\GitHub\pandaSDMX\pandasdmx\reader\sdmxml.py in _parse(self, elem, unwrap)
    404                 else:
    405                     # All other elements
--> 406                     result = [getattr(self, f'parse_{how}'.lower())(child)]
    407             except Reparse as r:
    408                 # Raise one level beyond the original to reparse <Parent><Ref>

~\Documents\GitHub\pandaSDMX\pandasdmx\reader\sdmxml.py in parse_structures(self, elem)
    908
    909     def parse_structures(self, elem):
--> 910         return self._parse(elem, unwrap=False)
    911
    912     def parse_organisation(self, elem):

~\Documents\GitHub\pandaSDMX\pandasdmx\reader\sdmxml.py in _parse(self, elem, unwrap)
    395                     # Element is a bare container for other elements; parse its
    396                     # children directly
--> 397                     result = list(chain(*self._parse(child, unwrap=False)
    398                                         .values()))
    399                 elif how == 'ref' or (len(child) == 1 and

~\Documents\GitHub\pandaSDMX\pandasdmx\reader\sdmxml.py in _parse(self, elem, unwrap)
    404                 else:
    405                     # All other elements
--> 406                     result = [getattr(self, f'parse_{how}'.lower())(child)]
    407             except Reparse as r:
    408                 # Raise one level beyond the original to reparse <Parent><Ref>

~\Documents\GitHub\pandaSDMX\pandasdmx\reader\sdmxml.py in parse_contentconstraint(self, elem)
   1229         role = elem.attrib.pop('type').lower()
   1230         elem.attrib['role'] = 'allowable' if role == 'allowed' else role
-> 1231         cc, values = self._named(ContentConstraint, elem)
   1232         cc.content.update(wrap(values.pop('constraintattachment')))
   1233         cc.data_content_region.append(values.pop('cuberegion', None))

~\Documents\GitHub\pandaSDMX\pandasdmx\reader\sdmxml.py in _named(self, cls, elem, **kwargs)
    526
    527         # Parse children
--> 528         values = self._parse(elem, **kwargs)
    529
    530         # Store the name, description and annotations

~\Documents\GitHub\pandaSDMX\pandasdmx\reader\sdmxml.py in _parse(self, elem, unwrap)
    404                 else:
    405                     # All other elements
--> 406                     result = [getattr(self, f'parse_{how}'.lower())(child)]
    407             except Reparse as r:
    408                 # Raise one level beyond the original to reparse <Parent><Ref>

~\Documents\GitHub\pandaSDMX\pandasdmx\reader\sdmxml.py in parse_cuberegion(self, elem)
   1237
   1238     def parse_cuberegion(self, elem):
-> 1239         values = self._parse(elem)
   1240         cr = CubeRegion(
   1241             included=elem.attrib.pop('include'),

~\Documents\GitHub\pandaSDMX\pandasdmx\reader\sdmxml.py in _parse(self, elem, unwrap)
    421             except Exception as e:
    422                 # Other exception, convert to XMLParseError
--> 423                 raise XMLParseError(self, child) from e
    424                 # NOTE to debug, use:
    425                 # raise e

XMLParseError: KeyError: 'conceptidentity'

Stack:
Structure
> Structures
> Constraints
> ContentConstraint
> CubeRegion

<com:Attribute xmlns:com="http://www.sdmx.org/resources/sdmxml/schemas/v2_1/common" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:mes="http://www.sdmx.org/resources/sdmxml/schemas/v2_1/message" xmlns:str="http://www.sdmx.org/resources/sdmxml/schemas/v2_1/structure" id="TIME_FORMAT">
  <com:Value>P1D</com:Value>
  <com:Value>P1M</com:Value>
  <com:Value>P1Y</com:Value>
  <com:Value>P3M</com:Value>
  <com:Value>P6M</com:Value>
  <com:Value>P7D</com:Value>
  <com:Value>PT1M</com:Value>
</com:Attribute>

(base) C:\Users\a3305\Documents\GitHub\pandaSDMX>activate sdmx09
Could not find conda environment: sdmx09
You can list all discoverable environments with `conda info --envs`.


(base) C:\Users\a3305\Documents\GitHub\pandaSDMX>activate pandasdmx09

(pandasdmx09) C:\Users\a3305\Documents\GitHub\pandaSDMX>cd ..
In [2]: from pandasdmx import Request


In [4]: imf=Request('imf_sdmxcentral')

In [5]: dsd_resp=imf.datastructure('ECOFIN_DSD')

In [6]:

In [6]: dsd_df = dsd_resp.write()

In [7]: dsd_df
Out[7]:
{'dataflow':                                                               name
 dataflow
 01R              Exchange Rates and International Reserves (01R...
 1DI              Coordinated Direct Investment Survey (CDIS) fo...
 1PI              Coordinated Portfolio Investment Survey (CPIS)...
 1SR                         M&B: Central Bank (1SR) for Collection
 2SR              M&B: Other Depository Corporations (2SR) for C...
 ...                                                            ...
 SOC                                              Socio-Demographic
 SPI                                Stock Market: Share Price Index
 UEM                                                   Unemployment
 WOE                                             Wages and Earnings
 ZZZ_TEST_ECOFIN                           Test Dataflow EcoFin DSD

 [75 rows x 1 columns],
 'conceptscheme':                                                                   name
 ECOFIN_CONCEPTS       NaN                      Concepts for EcoFin DSD
                       BASE_PER                             Base Period
                       COMMENT                                  Comment
                       COUNTERPART_AREA                Counterpart Area
                       DATA_DOMAIN                          Data Domain
 ...                                                                ...
 CROSS_DOMAIN_CONCEPTS UNIT_MULT                        Unit multiplier
                       USER_NEEDS                Relevance - user needs
                       USER_SAT          Relevance - user satisifaction
                       VALUATION                              Valuation
                       VIS_AREA              Counterpart reference area

 [145 rows x 1 columns],
 'codelist':                                                           name
 CL_DATADOMAIN  NaN                                 Data Domain
                01R   Exchange Rates and International Reserves
                1DI        Coordinated Direct Investment Survey
                1PI     Coordinated Portfolio Investment Survey
                1SR                           M&B: Central Bank
 ...                                                        ...
 CL_TIME_FORMAT P1Y                                      Annual
                P3M                                   Quarterly
                P6M                       Half-yearly, semester
                P7D                                      Weekly
                PT1M                                   Minutely

 [67303 rows x 1 columns]}

In [8]:  
```

**Comment by khaeru** [comment]
- See https://github.com/dr-leo/pandaSDMX/labels — the label *pre-1.0* has the description "Issues relating to 0.x releases". This is not the same as "issues that you would like fixed before 1.0 is released", so I suspect the label was applied by mistake. I will remove it; also from #164.
- You write "Should be fixed pre v1.0 release", but then assigned this to the 1.1 milestone. Which is it?

**Comment by dr-leo** [comment]
Some additional observations:

* I ran the Code on #151 rather than master.
* Issuing `bop_flow = imf.dataflow('BOP_BPM6')` does work contrary to requesting the datastructure directly which I find obscure.
* Inspecting the dimensions and Attributes described by the DSD, it turns out that 'TIME_FORMAT' occurs both as Dimension and Attribute. I suspect this is a bug, but haven't looked at the XML. I haven't checked whether this happens already in v0.9. If so, the key for a data request would be messed up which could be the reason for the failure described in #119.

**Comment by khaeru** [comment]
Fixed by #166.

I didn't get an answer to:
> You write "Should be fixed pre v1.0 release", but then assigned this to the 1.1 milestone. Which is it?

…but, moved this back to the 1.0 milestone since it's done and we still have not managed to release 1.0.

For the record…

> - Inspecting the dimensions and Attributes described by the DSD, it turns out that 'TIME_FORMAT' occurs both as Dimension and Attribute. I suspect this is a bug, but haven't looked at the XML.

This is not a violation of the IM. Dimensions and Attributes can have the same ID. In this case, both refer to the same TIME_FORMAT Concept and CodeList. In some IMF dataflows, this concept is expressed as an attribute; in others, as a dimension.

> - Issuing bop_flow = imf.dataflow('BOP_BPM6') does work contrary to requesting the datastructure directly which I find obscure.

This is because of the automatic adding of `?references=all` to the query. The BOP_BPM6 data flow uses ECOFIN_DSD, but its own ContentConstraint refers only to Dimensions:
```python
>>> import pandasdmx as sdmx
>>> IMF = sdmx.Request('IMF')
>>> IMF.dataflow('BOP_BPM6').constraint[0].data_content_region[0].member.keys()    
dict_keys([<Dimension: COUNTERPART_AREA>, <Dimension: DATA_DOMAIN>, <Dimension: INDICATOR>, <Dimension: TIME_FORMAT>])
```
On the other hand, asking for ECOFIN_DSD with references=all gives *all* the (hundreds of) data flows that are structured by it, and the associated constraint for every one of these data flows, such as:
```python
>>> IMF.dataflow('SPI').constraint[0].data_content_region[0].member.keys()                                                      
dict_keys([<Dimension: COUNTERPART_AREA>, <Dimension: DATA_DOMAIN>, <Dimension: INDICATOR>, <DataAttribute: TIME_FORMAT>])
```
So (prior to the fix) the parser failed. (This particular example was used in the PR.)

> - I ran the Code on #151 rather than master.

Comments on PR code should always go on the PR itself.

**Comment by dr-leo** [comment]
Similar error: 

```

In [10]:

In [10]: imf=Request('imf')
In [13]: dsd_resp=imf.datastructure('ECOFIN_DSD')
---------------------------------------------------------------------------
XMLParseError                             Traceback (most recent call last)
<ipython-input-13-633552b89f31> in <module>
----> 1 dsd_resp=imf.datastructure('ECOFIN_DSD')

~\Documents\GitHub\pandaSDMX\pandasdmx\api.py in get(self, resource_type, resource_id, tofile, use_cache, dry_run, **kwargs)
    430         # Parse the message, using any provided or auto-queried DSD
    431         msg = reader.read_message(response_content,
--> 432                                   dsd=kwargs.get('dsd', None))
    433
    434         # Store the HTTP response with the message

~\Documents\GitHub\pandaSDMX\pandasdmx\reader\sdmxml.py in read_message(self, source, dsd)
    287
    288         # Parse the tree
--> 289         values = self._parse(root)
    290
    291         # Instantiate the message object

~\Documents\GitHub\pandaSDMX\pandasdmx\reader\sdmxml.py in _parse(self, elem, unwrap)
    406                 else:
    407                     # All other elements
--> 408                     result = [getattr(self, f'parse_{how}'.lower())(child)]
    409             except Reparse as r:
    410                 # Raise one level beyond the original to reparse <Parent><Ref>

~\Documents\GitHub\pandaSDMX\pandasdmx\reader\sdmxml.py in parse_structures(self, elem)
    909
    910     def parse_structures(self, elem):
--> 911         return self._parse(elem, unwrap=False)
    912
    913     def parse_organisation(self, elem):

~\Documents\GitHub\pandaSDMX\pandasdmx\reader\sdmxml.py in _parse(self, elem, unwrap)
    397                     # Element is a bare container for other elements; parse its
    398                     # children directly
--> 399                     result = list(chain(*self._parse(child, unwrap=False)
    400                                         .values()))
    401                 elif how == 'ref' or (len(child) == 1 and

~\Documents\GitHub\pandaSDMX\pandasdmx\reader\sdmxml.py in _parse(self, elem, unwrap)
    406                 else:
    407                     # All other elements
--> 408                     result = [getattr(self, f'parse_{how}'.lower())(child)]
    409             except Reparse as r:
    410                 # Raise one level beyond the original to reparse <Parent><Ref>

~\Documents\GitHub\pandaSDMX\pandasdmx\reader\sdmxml.py in parse_contentconstraint(self, elem)
   1234         role = elem.attrib.pop('type').lower()
   1235         elem.attrib['role'] = 'allowable' if role == 'allowed' else role
-> 1236         cc, values = self._named(ContentConstraint, elem)
   1237         cc.content.update(wrap(values.pop('constraintattachment')))
   1238         cc.data_content_region.append(values.pop('cuberegion', None))

~\Documents\GitHub\pandaSDMX\pandasdmx\reader\sdmxml.py in _named(self, cls, elem, **kwargs)
    532
    533         # Parse children
--> 534         values = self._parse(elem, **kwargs)
    535
    536         # Store the name, description and annotations

~\Documents\GitHub\pandaSDMX\pandasdmx\reader\sdmxml.py in _parse(self, elem, unwrap)
    406                 else:
    407                     # All other elements
--> 408                     result = [getattr(self, f'parse_{how}'.lower())(child)]
    409             except Reparse as r:
    410                 # Raise one level beyond the original to reparse <Parent><Ref>

~\Documents\GitHub\pandaSDMX\pandasdmx\reader\sdmxml.py in parse_cuberegion(self, elem)
   1242
   1243     def parse_cuberegion(self, elem):
-> 1244         values = self._parse(elem, unwrap=False)
   1245         cr = CubeRegion(included=elem.attrib['include'])
   1246

~\Documents\GitHub\pandaSDMX\pandasdmx\reader\sdmxml.py in _parse(self, elem, unwrap)
    423             except Exception as e:
    424                 # Other exception, convert to XMLParseError
--> 425                 raise XMLParseError(self, child) from e
    426                 # NOTE to debug, use:
    427                 # raise e

XMLParseError: AttributeError: 'DataProvider' object has no attribute 'structure'

Stack:
Structure
> Structures
> Constraints
> ContentConstraint
> CubeRegion

<com:KeyValue xmlns:com="http://www.sdmx.org/resources/sdmxml/schemas/v2_1/common" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:mes="http://www.sdmx.org/resources/sdmxml/schemas/v2_1/message" xmlns:str="http://www.sdmx.org/resources/sdmxml/schemas/v2_1/structure" id="REF_AREA">
  <com:Value>US</com:Value>
</com:KeyValue>

```
